File: packages/skill-eval/src/config/env.ts
import fs from "node:fs";
import path from "node:path";
import { ENV_KEY_MAP } from "./registry.js";

function interpolateEnvVars(value: string): string {
	return value.replace(/\$\{(\w+)\}/g, (_, name) => process.env[name] ?? "");
}

export function resolveApiKey(providerName: string, explicitKey?: string): string | undefined {
	if (explicitKey) {
		const interpolated = interpolateEnvVars(explicitKey);
		if (interpolated) return interpolated;
	}
	const envNames = ENV_KEY_MAP[providerName] ?? [];
	for (const envName of envNames) {
		const val = process.env[envName];
		if (val) return val;
	}
	return undefined;
}

/**
 * Load .env files into process.env. Existing env vars take precedence.
 * For each starting directory, walks up to find the nearest .env file.
 */
export function loadDotenv(dirs: string[] = [process.cwd()]) {
	const loaded = new Set<string>();

	for (const startDir of dirs) {
		const envPath = findUp(".env", startDir);
		if (!envPath || loaded.has(envPath)) continue;
		loaded.add(envPath);
		parseEnvFile(envPath);
	}
}

function findUp(filename: string, startDir: string): string | null {
	let dir = path.resolve(startDir);
	while (true) {
		const candidate = path.join(dir, filename);
		if (fs.existsSync(candidate)) return candidate;
		const parent = path.dirname(dir);
		if (parent === dir) return null;
		dir = parent;
	}
}

function parseEnvFile(envPath: string) {
	let content: string;
	try {
		content = fs.readFileSync(envPath, "utf-8");
	} catch {
		return;
	}

	for (const line of content.split("\n")) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#")) continue;

		const eqIdx = trimmed.indexOf("=");
		if (eqIdx === -1) continue;

		const key = trimmed.slice(0, eqIdx).trim();
		let value = trimmed.slice(eqIdx + 1).trim();

		if (
			(value.startsWith('"') && value.endsWith('"')) ||
			(value.startsWith("'") && value.endsWith("'"))
		) {
			value = value.slice(1, -1);
		}

		if (key && process.env[key] === undefined) {
			process.env[key] = value;
		}
	}
}

File: packages/skillet-eval-python/src/skill_eval/config/env.py
from __future__ import annotations

import os
import re
from pathlib import Path

from dotenv import load_dotenv as _load_dotenv_file

from .registry import ENV_KEY_MAP

_ENV_INTERPOLATION = re.compile(r"\$\{(\w+)\}")


def interpolate_env_vars(value: str) -> str:
    def replacer(match: re.Match[str]) -> str:
        return os.environ.get(match.group(1), "")

    return _ENV_INTERPOLATION.sub(replacer, value)


def resolve_api_key(provider_name: str, explicit_key: str | None = None) -> str | None:
    if explicit_key:
        interpolated = interpolate_env_vars(explicit_key)
        if interpolated:
            return interpolated
    env_names = ENV_KEY_MAP.get(provider_name, [])
    for env_name in env_names:
        val = os.environ.get(env_name)
        if val:
            return val
    return None


def _find_up(filename: str, start_dir: str) -> Path | None:
    path = Path(start_dir).resolve()
    while True:
        candidate = path / filename
        if candidate.exists():
            return candidate
        parent = path.parent
        if parent == path:
            return None
        path = parent


def load_dotenv(dirs: list[str] | None = None) -> None:
    if dirs is None:
        dirs = [str(Path.cwd())]
    loaded: set[Path] = set()
    for start_dir in dirs:
        env_path = _find_up(".env", start_dir)
        if env_path is None or env_path in loaded:
            continue
        loaded.add(env_path)
        _load_dotenv_file(env_path)
